File: lang/python/src/fun/vac/algorithms/number_theory/GreatestCommonDivisor.py
def iterativeBinaryGCD(m: int, n: int) -> int:
    if m < 0:
        m = -m
    if n < 0:
        n = -n

    shift = 0

    while True:
        if m == n:
            return m << shift
        if m == 0:
            return n << shift
        if n == 0:
            return m << shift

        if (m & 1) == 0:
            m >>= 1
            if (n & 1) == 0:
                n >>= 1
                shift += 1
        else:
            if (n & 1) == 0:
                n >>= 1
            else:
                if m > n:
                    m = (m - n) >> 1
                else:
                    n = (n - m) >> 1

def recursiveBinaryGCD(m: int, n: int) -> int:
    if m < 0:
        m = -m
    if n < 0:
        n = -n

    if m == n:
        return m
    if m == 0:
        return n
    if n == 0:
        return m

    if (m & 1) == 0:
        if (n & 1) == 0:
            return recursiveBinaryGCD(m >> 1, n >> 1) << 1
        return recursiveBinaryGCD(m >> 1, n)
    if (n & 1) == 0:
        return recursiveBinaryGCD(m, n >> 1)
    if m > n:
        return recursiveBinaryGCD((m - n) >> 1, n)
    return recursiveBinaryGCD(m, (n - m) >> 1)

def iterativeEuclidean(m: int, n: int) -> int:
    if m < 0:
        m = -m
    if n < 0:
        n = -n

    while n != 0:
        m, n = n, m % n

    return m

def recursiveEuclidean(m: int, n: int) -> int:
    if m < 0:
        m = -m
    if n < 0:
        n = -n

    if n == 0:
        return m
    return recursiveEuclidean(n, m % n)

File: lang/ruby/src/fun/vac/algorithms/number_theory/GreatestCommonDivisor.rb
module GreatestCommonDivisor

    def self.iterativeBinaryGCD(m, n)
        if m.negative?
            m = -m
        end
        if n.negative?
            n = -n
        end

        shift = 0

        loop do
            if m == n
                return m << shift
            end
            if m.zero?
                return n << shift
            end
            if n.zero?
                return m << shift
            end

            if (m & 1).zero?
                m >>= 1
                if (n & 1).zero?
                    n >>= 1
                    shift += 1
                end
            else
                if (n & 1).zero?
                    n >>= 1
                else
                    if m > n
                        m = (m - n) >> 1
                    else
                        n = (n - m) >> 1
                    end
                end
            end
        end
    end

    def self.recursiveBinaryGCD(m, n)
        if m.negative?
            m = -m
        end
        if n.negative?
            n = -n
        end

        if m == n
            return m
        end
        if m.zero?
            return n
        end
        if n.zero?
            return m
        end

        if (m & 1).zero?
            if (n & 1).zero?
                return recursiveBinaryGCD(m >> 1, n >> 1) << 1
            end
            return recursiveBinaryGCD(m >> 1, n)
        end
        if (n & 1).zero?
            return recursiveBinaryGCD(m, n >> 1)
        end
        if m > n
            return recursiveBinaryGCD((m - n) >> 1, n)
        end
        recursiveBinaryGCD(m, (n - m) >> 1)
    end

    def self.iterativeEuclidean(m, n)
        if m.negative?
            m = -m
        end
        if n.negative?
            n = -n
        end

        until n.zero?
            m, n = n, m % n
        end

        m
    end

    def self.recursiveEuclidean(m, n)
        if m.negative?
            m = -m
        end
        if n.negative?
            n = -n
        end

        if n.zero?
            return m
        end
        recursiveEuclidean(n, m % n)
    end
end
